### DISCUSSION TITLE
Vehicle should unload during first visit of delivery location

### DISCUSSION THREAD

**Discussion by kazitanvirahsan** [google/or-tools#2845] [Routing (and legacy CP) questions]
Hai,

( Sorry guys I don't seem to delete the discussion and create a fresh one so I keep editing)

In my modeling, we have pickups and deliveries and vehicles can visit a delivery location multiple times. However,
some loads get picked up but not delivered during the first visit to the delivery location. It eventually gets delivered in the second visit to the delivery location.

Is there any way that I can able to limit pickup to delivery time duration so trucks delivery as soon as it visits the delivery location?

What I tried so far

```python
solver.Add(
  time_dimension.CumulVar(delivery_index) <=
  time_dimension.CumulVar(pickup_index) + 120)
```

Is this the right way to add a constraint to the solver? It seems to work with smaller datasets but takes time to return results with smaller jobs. There is a similar thread that mentions some problems with it.

I have another option use `SetPickupToDeliveryLimitFunctionForPair`.
https://github.com/google/or-tools/blob/86d4c543f717fa8716a9d66c25143660668bf825/ortools/constraint_solver/routing.h#L2767-L2779
Can anyone teach me how to use this function in python language?

Kazi

**Comment by Mizux** [comment]
Few questions:
- Each location can only be visited once so could you explain (code snippet?) what do you mean by second visit ?
- If you have some reload node you could add a constraint that upon arrival your delivery counter must be zero (i.e. before returning to depot to reload you must deliver all currently loaded goods)

**Comment by kazitanvirahsan** [comment]
Hai Mizuk,
    Thanks. Please look into the below screenshot I have attached.

[image: Screen Shot 2021-10-14 at 7.34.50 pm.png]

<img width="1394" alt="Screen Shot 2021-10-14 at 7 34 50 pm" src="https://user-images.githubusercontent.com/6487015/137291322-3a5cdb04-bac0-47d2-8174-de87270e3038.png">

Here truck's pick up load 10923444 on route index 13 and visits its
delivery location - SYDNEY CROSS DOCK HUB first time but not delivered yet.
The vehicle eventually unloads 10923444 on route index 13 when it visits
again SYDNEY CROSS DOCK HUB in the second visit.
We want to force the truck so that 10923444 can be delivered during its
first visit.

We do not have any node where trucks can reload at the moment. Vehicle
returns with some undelivered loads sometimes. but it would be good to know.

Kazi

**Comment by kazitanvirahsan** [comment]
Hai Mizuk,
    
I tried the following but I think it's not a very good solution as we don't know when the load will get delivered. 
```py
    solver.Add(
            travel_dimension.CumulVar(delivery_index) <=
            travel_dimension.CumulVar(pickup_index) + 180)
```
I was thinking if there is any way to implement penalties to consignment so higher penalties if unload on its second visit. Not sure.
Kazi

**Comment by kazitanvirahsan** [comment]
Hai @Mizux ,

    Can you please let me know if this is even possible?   

    1.  Add another node to the delivery location. We call it Exit Node of location.
    2. All the delivery nodes associated with that location must visit that node.
    3. Force to unload only those loads if delivery locations are the same (vehicle can have other loads to different delivery locations carrying. So not everything) in exit load.

Kind Regards
Kazi

**Reply by Mizux** [reply]
Not sure what you want to do, but if you want each vehicle to be able to visit delivery location only one time I would try to:
* Create one delivery "gate" `G_X_vY` location for each vehicle `Y` for each physical location `X`.
  * this node should be optional with 0 penalty, e.g. vehicle not visiting any nodes at location X should just drop this "gate" node
* All nodes at locations X have as `NextVar()`: any node at location X **and** nodes `G_X_v*`.
* Since upon departure from any node X vehicle will end up by visiting gate node which can be visited only once you guaranty that no vehicle can visit location X two time during its route.

ps: We also have a discord (link in the readme, see the  "chat" badge).
